src/contracts/core/StakeRootCompendium.sol

```diff
@@ -234,226 +202,155 @@ contract StakeRootCompendium is StakeRootCompendiumStorage {
+    function _setChargePerProof(uint64 _chargePerProofLinear, uint64 _chargePerProofConstant) internal {
+        _updateCumulativeCharges();
+        chargePerProofLinear = _chargePerProofLinear;
+        chargePerProofConstant = _chargePerProofConstant;
     }
 
-    // updates the deposit balance for the operator set and returns the penalty if the operator set has fallen below the minimum deposit balance
-    function _updateDepositBalanceInfo(
-        IAVSDirectory.OperatorSet memory operatorSet,
-        bool sendPenalty
-    ) internal returns (uint256) {
+    function _setProofInterval(uint32 proofInterval) internal {
         _updateCumulativeCharges();
-        (uint256 depositBalance, uint256 penalty) = getDepositBalance(operatorSet);
-        depositBalanceInfo[operatorSet.avs][operatorSet.operatorSetId].balance = depositBalance;
-        depositBalanceInfo[operatorSet.avs][operatorSet.operatorSetId].latestUpdateTime = uint32(block.timestamp);
-        // if the operatorSet has fallen below the minimum deposit balance, remove it from the stakeTree
-        if (penalty > 0) {
-            _removeOperatorSet(operatorSet);
-            if (sendPenalty) {
-                payable(msg.sender).transfer(penalty);
-            }
+        proofInterval = proofInterval;
+    }
+
+    function _updateCumulativeCharges() internal {
+        // todo comment on type casting
+        (uint96 _totalChargeLinear, uint96 _totalChargeConstant) = _cumulativeCharges();
+        if (_totalChargeLinear != totalChargeLinear || _totalChargeConstant != totalChargeConstant) {
+            totalChargeLinear = _totalChargeLinear;
+            totalChargeConstant = _totalChargeConstant;
+            blockNumberLastChargeUpdated = uint64(block.number);
         }
+    }
 
-        return penalty;
+    function _cumulativeCharges() internal view returns (uint96, uint96) {
+        uint96 proofEpochs = uint96((block.number - blockNumberLastChargeUpdated) / proofInterval);
+        return (totalChargeLinear + chargePerProofLinear * proofEpochs, totalChargeConstant + chargePerProofConstant * proofEpochs);
     }
 
-    // updates total strategies and the total charge per proof whenever an operator set's number of strategies changes
-    function _updateTotals(uint256 numStrategiesBefore, uint256 numStrategiesAfter) internal {
-        totalStrategies = totalStrategies - numStrategiesBefore + numStrategiesAfter;
-        totalChargeSnapshot.push(
-            uint32(block.timestamp), 
-            uint224(totalStrategies * linearChargePerProof + constantChargePerProof)
-        );
+    // updates the deposit balance for the operator set and returns the penalty if the operator set has fallen below the minimum deposit balance
+    function _updateDepositInfo(IAVSDirectory.OperatorSet memory operatorSet) internal returns (uint256) {
+        uint256 depositBalance = getDepositBalance(operatorSet);
```

Comment:
we should remove them from the operatorSet if they fall below the min balance here?

Reply:
we can, i reserved that logic where it is explicitly relevant eg processing withdrawals. . .

let me ponder, can put it back in, just thinking if there was a good reason i took it out